src/main/java/org/springframework/data/r2dbc/repository/query/PartTreeR2dbcQuery.java

```diff
@@ -0,0 +1,148 @@
+        ParameterMetadataProvider parameterMetadataProvider = new ParameterMetadataProvider(accessor, likeEscaper);
+        R2dbcQueryCreator queryCreator = new R2dbcQueryCreator(tree, dataAccessStrategy, entityMetadata,
+                parameterMetadataProvider);
+        String sql =  queryCreator.createQuery(getDynamicSort(accessor));
+        return new PartTreeBindableQuery(sql, accessor, parameterMetadataProvider);
+    }
+
+    private Sort getDynamicSort(RelationalParameterAccessor accessor) {
+        return parameters.potentiallySortsDynamically() ? accessor.getSort() : Sort.unsorted();
+    }
+
+    private static void validate(PartTree tree, RelationalParameters parameters, String methodName) {
+        int argCount = 0;
+        Iterable<Part> parts = () -> tree.stream().flatMap(Streamable::stream).iterator();
+        for (Part part : parts) {
+            int numberOfArguments = part.getNumberOfArguments();
+            for (int i = 0; i < numberOfArguments; i++) {
+                throwExceptionOnArgumentMismatch(methodName, part, parameters, argCount);
+                argCount++;
+            }
+        }
+    }
+
+    private static void throwExceptionOnArgumentMismatch(String methodName,
+                                                         Part part,
+                                                         RelationalParameters parameters,
+                                                         int index) {
+        Part.Type type = part.getType();
+        String property = part.getProperty().toDotPath();
+
+        if (!parameters.getBindableParameters().hasParameterAt(index)) {
+            String msgTemplate = "Method %s expects at least %d arguments but only found %d. " +
+                    "This leaves an operator of type %s for property %s unbound.";
+            String formattedMsg = String.format(msgTemplate, methodName, index + 1, index, type.name(), property);
+            throw new IllegalStateException(formattedMsg);
+        }
+
+        RelationalParameters.RelationalParameter parameter = parameters.getBindableParameter(index);
+        if (expectsCollection(type) && !parameterIsCollectionLike(parameter)) {
+            String message = wrongParameterTypeMessage(methodName, property, type, "Collection", parameter);
+            throw new IllegalStateException(message);
+        } else if (!expectsCollection(type) && !parameterIsScalarLike(parameter)) {
+            String message = wrongParameterTypeMessage(methodName, property, type, "scalar", parameter);
+            throw new IllegalStateException(message);
+        }
+    }
+
+    private static boolean expectsCollection(Part.Type type) {
+        return type == Part.Type.IN || type == Part.Type.NOT_IN;
+    }
+
+    private static boolean parameterIsCollectionLike(RelationalParameters.RelationalParameter parameter) {
+        return Iterable.class.isAssignableFrom(parameter.getType()) || parameter.getType().isArray();
```

Comment:
We learned the hard way that we shouldn't use `Iterable` here but `Collection`. Also we should check first if the type of the parameter matches the type of the property. See https://jira.spring.io/browse/DATAJPA-1682